### PR TITLE
WIP: Logger

### PR DESCRIPTION
**Pull request by Setogit**
cc @bajtos @raymondfeng @ritch @superkhau

DO NOT MERGE

Logger component to packages/logger
Client code to packages/example-codehub

### CODE REVIEW

File: packages/authentication/index.d.ts [hunk 1]
```diff
@@ -1,6 +1 @@
-// Copyright IBM Corp. 2013,2017. All Rights Reserved.
-// Node module: @loopback/authentication
-// This file is licensed under the MIT License.
-// License text available at https://opensource.org/licenses/MIT
-
-export * from './lib';
+export * from './src';
```
Comment (bajtos) [inline packages/authentication/index.d.ts]: Please revert all changes in all `index.d.ts` and `index.js` files. I don't mind discussing what's the right content to place there (e.g. I agree we should be enabling strict mode), but that should be done in a standalone pull request.

File: packages/example-codehub/src/codehub-application.ts [hunk 2]
```diff
@@ -3,12 +3,46 @@
...
+  async handle(req: any, res: any) {
+    if (!this.httpLogger) {
+      this.httpLogger = pinoHttp({
+        logger: this.loggerProvider,
+        genReqId: function(req: any) { return req.id },
+      });
+      console.log('~~~ PinoHttp logger initialized');
+    }
+    this.httpLogger(req, res);
+    this.defaultHandle(req, res);
```
Comment (bajtos) [inline packages/example-codehub/src/codehub-application.ts]: Why do we need `this.defaultHandle`? Cannot we simply call `super(request, response)` instead? 

See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/super

File: packages/example-codehub/src/codehub-application.ts [hunk 3]
```diff
@@ -3,12 +3,46 @@
...
+  @inject(loggerProviderKey)
+  private loggerProvider: Provider<Logger>;
+  private httpLogger: any;
+  constructor(a: any, b: any, c: any, d: any) {
+    super(a, b, c, d);
+    this.defaultHandle = super.handle;
+  }
+  async handle(req: any, res: any) {
+    if (!this.httpLogger) {
+      this.httpLogger = pinoHttp({
```
Comment (bajtos) [inline packages/example-codehub/src/codehub-application.ts]: While this is functional, it's not following our design intentions for LoopBack next. See the newly added section https://github.com/strongloop/loopback-next/wiki/Writing-Components#modifying-request-handling-logic for more details.

IMO, the component should be exporting a provider for an action that can be used by user/application sequence. Something along the following lines:

```ts
import {SetupLogger} from '@loopback/logger';

class AppSequence extends DefaultSequence {
  constructor(
    @inject('sequence.actions.findRoute') protected findRoute: FindRoute,
    @inject('sequence.actions.invokeMethod') protected invoke: InvokeMethod,
    @inject('sequence.actions.send') public send: Send,
    @inject('sequence.actions.reject') public reject: Reject,
    @inject('logger.actions.setupLogger') public setupLogger: SetupLogger,
  ) {}

  async handle(req: ParsedRequest, res: ServerResponse) {
    this.setupLogger(req);
    await super(req, res);
  }
}
```

File: packages/example-codehub/src/codehub-application.ts [hunk 4]
```diff
@@ -3,12 +3,46 @@
...
+        logger: this.loggerProvider,
+        genReqId: function(req: any) { return req.id },
```
Comment (bajtos) [inline packages/example-codehub/src/codehub-application.ts]: There should be a well-defined binding key for obtaining request correlation id, I think there will be more components that will need this value. The last example in https://github.com/strongloop/loopback-next/wiki/Writing-Components#providers is showing how to implement such provider.

```ts
// somewhere in app setup
app.bind('request-correlation-id').toProvider(CorrelationIdProvider);
```

File: packages/example-codehub/src/codehub-application.ts (hunk 2 above)
Comment (bajtos) [inline packages/example-codehub/src/codehub-application.ts]: Also `defaultHandle` is returning a promise, you need to forward it up the caller chain. See my "class AppSequence extends DefaultSequence" example in a comment above.

File: packages/example-codehub/src/codehub-application.ts [hunk 5]
```diff
@@ -3,12 +3,46 @@
...
+  }
+}
+
 export class CodeHubApplication extends Application {
   constructor() {
-    super();
+    super({
+      components: [PinoLoggerComponent],
+      sequence: PinoHttpLoggerSequence,
+      // components: [WinstonLoggerComponent],
```
Comment (bajtos) [inline packages/example-codehub/src/codehub-application.ts]: I don't think we should be supporting more than a single log provider. If people want to use Winston, then they should write their own component. Our job is to select the provider we think is best for 80% of users, to make it easier for them to start and reduce the number of decisions they have to make when starting a new project.

File: packages/logger/README.md [hunk 6]
```diff
@@ -0,0 +1,4 @@
+# @loopback/logger
+
+A LoopBack component for logging.
+
```
Comment (bajtos) [inline packages/logger/README.md]: This will obviously need more content, see https://loopback.io/doc/en/contrib/README-guidelines.html

Let's leave that for the end though, we need to work out the design and implementation first.

File: packages/logger/index.ts [hunk 7]
```diff
@@ -0,0 +1,7 @@
+// Copyright IBM Corp. 2013,2017. All Rights Reserved.
+// Node module: @loopback/logger
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+// NOTE(bajtos) This file is used by VSCode/TypeScriptServer at dev time only
+export * from './src';
```
Comment (bajtos) [inline packages/logger/index.ts]: I think you will need to add `packages/logger/index.js` and `packages/logger/index.d.ts` with the same content as we have in other projects.

File: packages/logger/src/keys.ts [hunk 8]
```diff
@@ -0,0 +1,10 @@
+// Copyright IBM Corp. 2017. All Rights Reserved.
+// Node module: @loopback/logger
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+export namespace BindingKeys {
+  export namespace System {
+    export const LOGGER_PROVIDER: string = 'system.logger_provider';
```
Comment (bajtos) [inline packages/logger/src/keys.ts]: Let's use the component name as the prefix, i.e. `logger.`. If we end up exporting a sequence action as I mentioned in one of my comments above, then I think this should become:

```ts
export namespace Logger {
  export namespace SequenceActions {
    export const SETUP_LOGGER  = 'logger.sequence.actions.setupLogger';
  }
}
```

File: packages/authentication/index.d.ts (hunk 1 above)
Comment (Setogit) [inline packages/authentication/index.d.ts]: reverted the changes made unintentionally to authentication, context, and core.  Sorry!

File: packages/example-codehub/src/codehub-application.ts (hunk 3 above)
Comment (Setogit) [inline packages/example-codehub/src/codehub-application.ts]: done

File: packages/example-codehub/src/codehub-application.ts (hunk 2 above)
Comment (Setogit) [inline packages/example-codehub/src/codehub-application.ts]: got it. done.

File: packages/example-codehub/src/codehub-application.ts (hunk 5 above)
Comment (Setogit) [inline packages/example-codehub/src/codehub-application.ts]: agreed.  removed WinstonLogger implementation.

File: packages/logger/README.md (hunk 6 above)
Comment (Setogit) [inline packages/logger/README.md]: certainly.

File: packages/logger/index.ts (hunk 7 above)
Comment (Setogit) [inline packages/logger/index.ts]: added

File: packages/logger/src/keys.ts (hunk 8 above)
Comment (Setogit) [inline packages/logger/src/keys.ts]: done

File: packages/example-codehub/src/codehub-application.ts (hunk 4 above)
Comment (Setogit) [inline packages/example-codehub/src/codehub-application.ts]: the `genReqId` line is not required for pino-http.  remove the line for now to keep things simpler.

**Comment by Setogit** [comment]
Incorporated all comments except that:
- Kept defaultHandle because super calls are not permitted outside constructors or in nested functions inside constructors.
- Remove genReqId to keep it simple.  It is not required for pino-http.
- Will revisit README.md after the code cleanup.

File: packages/example-codehub/src/codehub-application.ts [hunk 9]
```diff
@@ -3,12 +3,42 @@
...
+  ParsedRequest, Reject, Send, inject} from '@loopback/core';
+import {Logger, PinoLoggerComponent} from '@loopback/logger';
 import {UserController, HealthController} from './controllers';
+import {ServerResponse} from 'http';
+
+const setupLoggerKey = Logger.SequenceActions.SETUP_LOGGER;
+
+class AppSequence extends DefaultSequence {
+  private defaultHandle: Function;
+  private httpLogger: any;
```
Comment (bajtos) [inline packages/example-codehub/src/codehub-application.ts]: We are trying to avoid using `any`. What is the the actual type of a logger function? We should use that type.

File: packages/example-codehub/src/codehub-application.ts [hunk 10]
```diff
@@ -3,12 +3,42 @@
...
+  constructor(
+    @inject('sequence.actions.findRoute') protected findRoute: FindRoute,
+    @inject('sequence.actions.invokeMethod') protected invoke: InvokeMethod,
+    @inject('sequence.actions.send') public send: Send,
+    @inject('sequence.actions.reject') public reject: Reject,
+    @inject(setupLoggerKey) public setupLogger: any,
```
Comment (bajtos) [inline packages/example-codehub/src/codehub-application.ts]: This is weird. You are making `setupLogger` a public property (`this.setupLogger` or `sequence.setupLogger`) and also storing it in `httpLogger` property too. I think `httpLogger` is not needed here.

File: packages/example-codehub/src/codehub-application.ts [hunk 11]
```diff
@@ -3,12 +3,42 @@
...
+  ) {
+    super(findRoute, invoke, send, reject);
+    this.httpLogger = setupLogger;
+    this.defaultHandle = super.handle;
+  }
+
+  async handle(req: ParsedRequest, res: ServerResponse) {
+    this.httpLogger(req, res);
+    await this.defaultHandle(req, res);
```
Comment (bajtos) [inline packages/example-codehub/src/codehub-application.ts]: > because tsc complains:
>
> - Constructors for derived classes must contain a 'super' call.
> - Super calls are not permitted outside constructors or in nested functions inside constructors.

Sorry, my mistake. I think you need to call `super.handle(req, res)`. Will that work?

File: packages/logger/index.d.ts [hunk 12]
```diff
@@ -0,0 +1 @@
+export * from './src';
```
Comment (bajtos) [inline packages/logger/index.d.ts]: I believe other packages are exporting type definitions from `./lib`. IIRC, we are not shipping `src` in npm packages.

File: packages/logger/index.js [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+// Copyright IBM Corp. 2013,2017. All Rights Reserved.
+// Node module: @loopback/logger
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+function __export(m) {
+    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
+}
+Object.defineProperty(exports, "__esModule", { value: true });
+// NOTE(bajtos) This file is used by VSCode/TypeScriptServer at dev time only
+__export(require("./src"));
```
Comment (bajtos) [inline packages/logger/index.js]: Please use the same contents of `index.js` as other packages are using. We need to switch between `lib` and `lib6` depending on the version of Node.js that we are running. See e.g. here: [packages/core/index.js](https://github.com/strongloop/loopback-next/blob/master/packages/core/index.js). Feel free to add missing `"use strict";` header though.

File: packages/logger/package.json [hunk 14]
```diff
@@ -0,0 +1,54 @@
+{
+  "name": "@loopback/logger",
+  "version": "1.0.0",
```
Comment (bajtos) [inline packages/logger/package.json]: Let's reserve `1.0.0` version for GA please. I am proposing to use `0.1.0` instead.

File: packages/logger/package.json [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+  "description": "A LoopBack component for logger support.",
+  "scripts": {
+    "acceptance": "mocha --opts ../../test/mocha.opts 'test/acceptance/**/*.ts'",
+    "build": "npm run build:lib && npm run build:lib6",
+    "build:lib": "node ../../bin/compile-package es2017",
+    "build:lib6": "node ../../bin/compile-package es2015",
+    "clean": "rm -rf loopback-logger*.tgz lib* package",
+    "integration": "mocha --opts ../../test/mocha.opts 'test/integration/**/*.ts'",
+    "prepare": "npm run build",
+    "pretest": "npm run build",
```
Comment (bajtos) [inline packages/logger/package.json]: This should be `npm run build:current` instead. I think we may have changed npm scripts since you started the work on this logger component. Please check e.g. [packages/core/package.json](https://github.com/strongloop/loopback-next/blob/021050065007937b8935f086fd4a4b230eb9396d/packages/core/package.json) for the correct content.

File: packages/logger/package.json [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+  },
+  "author": "IBM",
+  "license": "MIT",
+  "dependencies": {
+    "@loopback/context": "^4.0.0-alpha.9",
+    "@loopback/core": "^4.0.0-alpha.9",
+    "@types/pino": "^3.0.1",
+    "@types/shot": "^3.4.0",
+    "@types/sinon": "^2.3.3",
+    "@types/supertest": "^2.0.2",
```
Comment (bajtos) [inline packages/logger/package.json]: shot, sinon, supertest in regular (production) dependencies - that does not look right to me. Could you please review the dependencies and remove all unused ones?

File: packages/logger/package.json [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+    "pino": "^4.7.1",
+    "pino-http": "^2.6.1",
+    "shot": "^3.4.2",
+    "sinon": "^2.3.8",
+    "supertest": "^3.0.0",
+    "winston": "^2.3.1"
+  },
+  "devDependencies": {
+    "@loopback/openapi-spec-builder": "^4.0.0-alpha.4",
```
Comment (bajtos) [inline packages/logger/package.json]: I think this is not used either.

File: packages/logger/src/index.ts [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+const setupLoggerKey = Logger.SequenceActions.SETUP_LOGGER;
+const loggerProviderKey = Logger.System.LOGGER_PROVIDER;
+
+export {Logger} from './keys';
+export {SimpleLogger} from './providers/LoggerProvider';
+
+export class PinoLoggerComponent implements Component {
+  providers = {
+    // PinoLoggerProvider is bound to the key [loggerProviderKey]
+    // in the client's application context.
```
Comment (bajtos) [inline packages/logger/src/index.ts]: To be honest, I see very little value in this comment - it's just describing how components work.

The more interesting question is what is PinoLoggerProvider good for, where would it people use? This is not clear to me (I haven't read the tests yet), a comment explaining _that_ would be helpful.

The same comment applies to setupLoggerKey. I know what's that for because it's based on our past discussion, but people reading this code may not know.

File: packages/logger/src/keys.ts [hunk 19]
```diff
@@ -0,0 +1,13 @@
...
+// Node module: @loopback/logger
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+export namespace Logger {
+  export namespace SequenceActions {
+    export const SETUP_LOGGER  = 'logger.sequenceActions.setupLogger';
+  }
+  export namespace System {
+    export const LOGGER_PROVIDER  = 'logger.system.loggerProvider';
```
Comment (bajtos) [inline packages/logger/src/keys.ts]: 1) IMO, the term "Provider" is redundant here. A provider is a way how to register a value provider with our IoC container (`Context`). The actual value injected won't be a provider, but the logger instance returned by the provider. In that sense, the key name may be even misleading.

Here is what I am proposing instead:

```ts
export const LOGGER = 'logger.system.logger';
```

2) Could you please elaborate little bit more on the "system" name - why should the binding key include "system" in the name? IMO, there is nothing system-wide or system-level in LoopBack Next, so it does not make sense to me to include "system" in the binding key. Especially when the binding is contributed by an optional component. It's perfectly possible that I am missing something, that's why I'd like to hear your thoughts.

File: packages/logger/src/providers/LoggerProvider.ts [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+import {Logger} from '../keys';
+import {Provider} from '../../../context';
+import {inject} from '../../../core';
+
+const loggerProviderKey = Logger.System.LOGGER_PROVIDER;
+
+/**
+ * @description setup Pino HTTP logger
+ */
+export class SetupPinoHttpLogger implements Provider<any> {
```
Comment (bajtos) [inline packages/logger/src/providers/LoggerProvider.ts]: Let's leave "pino" from class names please. The fact that we are using pino is an implementation detail IMO. Even the component is called `@loopback/logger`, not `@loopback/pino` or `@loopback/pino-logger`.

File: packages/example-codehub/src/codehub-application.ts (hunk 2 above)
Comment (Setogit) [inline packages/example-codehub/src/codehub-application.ts]: Now, awaiting super.handle as suggested.

File: packages/example-codehub/src/codehub-application.ts (hunk 10 above)
Comment (Setogit) [inline packages/example-codehub/src/codehub-application.ts]: Removed it and awaited super.handle.

File: packages/logger/package.json (hunk 17 above)
Comment (Setogit) [inline packages/logger/package.json]: removed

File: packages/logger/package.json (hunk 16 above)
Comment (Setogit) [inline packages/logger/package.json]: removed them.

File: packages/logger/package.json (hunk 15 above)
Comment (Setogit) [inline packages/logger/package.json]: updated

File: packages/logger/src/index.ts (hunk 18 above)
Comment (Setogit) [inline packages/logger/src/index.ts]: removed the comment

File: packages/logger/src/keys.ts (hunk 19 above)
Comment (Setogit) [inline packages/logger/src/keys.ts]: The reason I tentatively used system.logger is that I was thinking there might be two or more loggers simultaneously and this one is so called `default logger`, but it was long time ago and my understanding/assumption at that time was less mature.
```
export namespace Logger {
  export const SIMPLE_LOGGER  = 'logger.simpleLogger';
```

File: packages/logger/src/providers/LoggerProvider.ts (hunk 20 above)
Comment (Setogit) [inline packages/logger/src/providers/LoggerProvider.ts]: got rid of all `pino` mention.

File: packages/logger/package.json (hunk 14 above)
Comment (Setogit) [inline packages/logger/package.json]: changed to 0.1.0

File: packages/logger/index.js (hunk 13 above)
Comment (Setogit) [inline packages/logger/index.js]: looking.  `index.js` file is created in the translilation step.

File: packages/logger/index.d.ts (hunk 12 above)
Comment (Setogit) [inline packages/logger/index.d.ts]: looking. index.js file is created in the translilation step.

File: packages/example-codehub/src/codehub-application.ts (hunk 11 above)
Comment (Setogit) [inline packages/example-codehub/src/codehub-application.ts]: yes, it works and I believe it was my mistake, sorry.

**Comment by Setogit** [comment]
index.js and index.d.ts are created by `tsc`  Looking...  Incorporated all the other comments.

**Comment by bajtos** [comment]
> index.js and index.d.ts are created by tsc Looking... Incorporated all the other comments.

How are you running `tsc`? The default `tsconfig.json` in project root includes all `*.ts` files in all packages, including "dummy" `index.ts` files in package roots. I think that if you run `tsc` on its own, then it will indeed rewrite `index.js` files.

However, our build system (`npm run build`) is using different tsconfig files (`packages/*/tsconfig.build.json`) which are configuring tsc to compile only files in `src` folder.

**Comment by bajtos** [comment]
@Setogit thank you for the update. Let's step back from the implementation details a bit and look at how this logger should be used, so that you can write acceptance tests and use them to guide the implementation.

From my POV, as a developer writing a LoopBack application, I want to access the logger in several different places:

At bootstrap/startup time, I want to log messages using app-level logger.

```ts
app.logger.info({url}, `Listening at ${url}`);
```

In my controllers, I want to log messages using request-level logger.

```ts
class MyController {
  constructor(@inject(...) protected logger: Logger) {}
  
  @get('/greet')
  @param.query.string('name')
  greet(name: string) {
    logger.info('Hello %s', name);
    return `Hello ${name}`
  }
}
```

Considering that pino-http is exposing only a single `log` function that's already bound to a configurable log level, it may make more sense to do the same in loopback-next too (or we can support both flavours).

```ts
class MyController {
  constructor(@inject(...) protected log: Log) {}
  
  @get('/greet')
  @param.query.string('name')
  greet(name: string) {
    log('Hello %s', name);
    return `Hello ${name}`
  }
}
```

Persistence-related code (think of Models and Repositories) can be run both at both startup time (e.g. from a boot script seeding the database with initial data) and at request time. It would be great if there was a way how to configure `Context` to inject app-wide or request-level logger depending on how and when the model/repository was instantiated. This way the code depending on the logger can stay decoupled of the particular logger implementation (app-wide, per-request).

```ts
class ShoppingCart {
  constructor(
    @inject(...) protected shippingCalculator: ShippingCostService,
    @inject(...) protected logger: Logger,
  ) {}

  async updateShippingFee() {
    this.logger.debug({id: this.id}, 'Calculating shipping fee.');
    const cost = await shippingCalculator.calculate(this.address);
    this.logger.debug({id: this.id, address: this.address, cost});
    // ...
  }
}
```

Last but not least, the logger should make it easy to log request/response related information, e.g. how long did it take to handle an incoming request. This is a feature provided by pino-http and it is the reason why we need a custom step in the Sequence IIUC.

These are the most important use cases I can think of. @Setogit What are your thoughts on this? Is there anything to add, remove or change?

cc @superkhau @ritch

File: packages/logger/README.md [hunk 21]
```diff
@@ -1,4 +1,320 @@
...
+
+
+# Foundation
+
+Here, I'm going to summarize a few key concepts and a list of tools I've found very useful.  With that, I got a big picture without getting bogged down to the details of each technology; and, at the same time, learned a way to go back the the details when needed.
+
+LoopBack Next is built on TypeScript and OpenAPI.  TypeScript is build on ECMAScript 6.  Quickly browsing those technologies from the ground up help solidify our steps. 
+
+### [ECMAScript 6](https://en.wikipedia.org/wiki/ECMAScript#6th_Edition_-_ECMAScript_2015)
```
Comment (bajtos) [inline packages/logger/README.md]: In fact, we are using ECMAScript 2017, which is an even newer version of the standard. (ES6 is an alias for ECMAScript 2015.)

The most notable language features we are using from ES2017 that are not in ES6:
 - `async`/`await`, a.k.a "async functions" (see https://tc39.github.io/ecmascript-asyncawait/)
 - trailing commas in function parameter lists and calls (see https://github.com/tc39/proposal-trailing-function-commas)

It would be great to mention at least "async functions" in the "Foundation" part, as we are using them pretty much everywhere in our codebase (and examples).

File: packages/logger/README.md [hunk 22]
```diff
@@ -1,4 +1,320 @@
...
+2. Declaration Files -- https://basarat.gitbooks.io/typescript/docs/types/ambient/d.ts.html
+3. Decorators -- https://www.typescriptlang.org/docs/handbook/decorators.html
+
+When you need to `npm install async` don't forget to `npm install -S "@types/async"
+
+### [OpenAPI](https://www.openapis.org/)
+
+1. https://editor.swagger.io
+
+Note (LoopBack Next specific): In a LoopBack Next app, typically you don't put the header portion of the api spec(see below) because LoopBack Next adds it when internalizing it.  The swagger editor will complain without the header portion.
```
Comment (bajtos) [inline packages/logger/README.md]: Let me clarify: the header portion of the api spec can be omitted when specifying API of a controller via `@api` decorator.

OTOH, `app.api()` requires a valid OpenAPI spec document, including the headers. This is checked only at compile time by TypeScript, we don't assert that on runtime.

If the app does not call `app.api()` to define header fields like "info", then we add some reasonable defaults to make the spec valid (see `createEmptyApiSpec`).

File: packages/logger/README.md [hunk 23]
```diff
@@ -1,4 +1,320 @@
...
+Also note that LoopBack Next provides `validateApiSpec` in the testlab so that you can validate it in `npm test` of your app.  Isn't that cool?
+
+
+
+# LoopBack Next
+
+We're going to examine three key concepts: controller & apiSpec, components & providers, and custom sequence.  Working sample codes are available in several places, one of which is [Hello World](https://github.com/strongloop/loopback-next-hello-world).  I'm going to focus on specific conceptual roadblocks I stumbled on.
+
+Skeleton of the client code looks like this:
+```
```
Comment (bajtos) [inline packages/logger/README.md]: Please specify the language as here - add `js` or `ts` right after the three backticks. Similarly in other code examples below.

File: packages/logger/README.md [hunk 24]
```diff
@@ -1,4 +1,320 @@
...
+            "description": "Returns a hello world with your (optional) name."
+          }
+        }
+      }
+    }
+  }
+}
+```
+@api is a [class decorator](https://www.typescriptlang.org/docs/handbook/decorators.html).  The class decorator,
+`api` is a function that is called when the class, HelloWorldController is instantiated.  In this case, `api` registers
```
Comment (bajtos) [inline packages/logger/README.md]: The decorator is a function that is called when the class is defined (only once), not every time it's instantiated.

File: packages/logger/README.md [hunk 25]
```diff
@@ -1,4 +1,320 @@
...
+the API specification to the application.  The decorator site describes some relatively complex concept, but we can just read the class decorator section to get a high-level understanding of decorator.  The concept is the same for other types of decorators.  That's enough.
+
+In the above sample code, @inject is a property decorator defined by LoopBack Next.  It's a very important LoopBack Next concept.  It works this way: @inject acquires the value bound to the key: `defaultName` in the application's context, then assigns the value to `name` property when the class `HelloWorldControlle` is instantiated.
+
+Please note that we introduced another new term, `Context`.  Context is simply a memory space where LoopBack Next maintains key-value pairs.  There are two types of context: application context and request context.  In this blog, we deal with only application context.
+
+The LoopBack Next specific decorator, @inject can be used as a parameter decorator e.g.:
```
Comment (bajtos) [inline packages/logger/README.md]: Yes and no. While decorators can be used for parameters in general, the `@inject` decorator is supported only for class constructor parameters.

```ts
class MyClass {
  constructor(
    @inject('defaultName') name: string
  } {}
}
```

File: packages/logger/README.md [hunk 26]
```diff
@@ -1,4 +1,320 @@
...
+```
+In this case, the argument `name` is replaced with the value bound to the key `defaultName`.  @inject acquires the value bound to the `defaultName` key in the application's context, then assigns the value to `name` argument when the helloWorld method is invoked.
+
+Who binds `defaultName` key to the value?  It depends.  You can do that by app.bind(`defaultName`).to('Ted Johnson').  You can `get` the value string by app.getSync('defaultName').
+
+More important usage of @inject is discussed in `Component and Provider` section.
+
+### Components and Providers
+
+Component defines a functionality.  LoopBack Next application can be viewed as a collection of components.  In a component, one or more providers implement the functionality.  In the Logger provider example below, (a) 'logger' key is  bound to the provider instance when Application is instantiated, and (b) the client is acquiring the logger instance by app.getSync('logger') and using the logger to display the message, 'My application has started.'.
```
Comment (bajtos) [inline packages/logger/README.md]: `app.get('logger')` please

File: packages/logger/README.md [hunk 27]
```diff
@@ -1,4 +1,320 @@
...
+
+Client:
+```
+const app = new Application({
+  components: [AuthenticationComponent, LoggerComponent],
+});
+
+const logger = app.getSync('logger');
```
Comment (bajtos) [inline packages/logger/README.md]:
```js
const logger = await app.get('logger');
```

File: packages/logger/README.md [hunk 28]
```diff
@@ -1,4 +1,320 @@
...
+  providers = {
+    [key]: LoggerProvider,  // [key] is a computed property name
+  }
+}
+```
+  Note: What's computed property name ?  See http://es6-features.org/#ComputedPropertyNames
+```
+import {Provider} from '@loopback/context';
+
+export class LoggerProvider implements Provider<SimpleLogger> {
```
Comment (bajtos) [inline packages/logger/README.md]: Let's use `Logger` instead of `SimpleLogger`. Using "Simple" as a name prefix is a sort of an anti-pattern: "simple" does not tell us any new information about the class behavior and how does it differ from other "complex?" loggers.

File: packages/logger/README.md [hunk 29]
```diff
@@ -1,4 +1,320 @@
...
+Usually, the default sequence implemented by LoopBack Next core works for many applications.  In those cases, there is nothing the application needs to do.  To add a piece of new functional element to your application, you will implement the entire sequence in your application.
+
+Client:
+```
+const app = new Application({
+  components: [AuthenticationComponent, LoggerComponent],
+});
+app.sequence(MySequence);
+
+const logger = app.getSync('logger');
```
Comment (bajtos) [inline packages/logger/README.md]:
```js
const logger = await app.get('logger');
```

Is it actually needed to repeat these two lines in this example?

File: packages/logger/README.md [hunk 30]
```diff
@@ -1,4 +1,320 @@
...
+Once attached, the loaded API spec can be accessed by `app.getApiSpec()`
+
+LoopBack Next provides the OpewnAPI spec validator, validateApiSpec as part of the lestlab.
+
+```
+const validateApiSpec = require('@loopback/testlab').validateApiSpec;
+
+describe('validate Api Spec', () => {
+  let app = createApp(); // instantiates the app, its controllers, and the API spec.
+  let apiSpec = app.getApiSpec();
```
Comment (bajtos) [inline packages/logger/README.md]: Test setup should go to a `before` or `beforeEach` hook.

Also test names should read as English sentence, where `it()` stands for the title from `describe`. In this code, we have two sentences: "it apisSpec is valid" (short test name) and "validate Api Spec apisSpec is valid" (full test name), none of read well to me.

```js
describe('Application\'s Api Spec', () => {
  let app, apiSpec;
  before(() => {
    app = createApp();
    apiSpec = app.getApiSpec();
  });

  it('is valid', async () => {
    await validateApiSpec(apiSpec);
  });
});
```

File: packages/logger/README.md [hunk 31]
```diff
@@ -1,4 +1,320 @@
...
+    await validateApiSpec(apiSpec);
+  });
+});
+
+```
+
+### How to debugt API Spec
+
+[The Swagger Editor](https://editor.swagger.io) is useful to interactively debug your OpenAPI specs.
+You can start with JSON or YAML.  Just cut and paste the helloworld API spec (below) into the swagger editor window.  It will ask you to use JSON or covert it to YAML and build the API spec in YAML.
```
Comment (bajtos) [inline packages/logger/README.md]: Maybe we can tell people where to find the API Spec document describing their application? Assuming the app is listening on the port 3000, the spec can be downloaded from `http://localhost:3000/openapi.json`.

File: packages/logger/README.md [hunk 32]
```diff
@@ -1,4 +1,320 @@
...
+            "description": "Returns a hello world with your (optional) name."
+          }
+        }
+      }
+    }
+  }
+}
+```
+
+### LoopBack Next globalization, heh?
```
Comment (bajtos) [inline packages/logger/README.md]: Ha, globalization is a good point that we have not considered yet. I think we should be extracting strings from the original TypeScript sources, what do you think, @Setogit?

Let's discuss globalization here: https://github.com/strongloop/loopback-next/issues/493

Important: globalization is out of scope of the initial beta release, but I think it should be a must-have for GA.

File: packages/logger/README.md (hunk 21 above)
Comment (Setogit) [inline packages/logger/README.md]: ECMAScript 2017  -- fixed
async/await -- added
trailing commas in function parameter lists and calls -- skipped

File: packages/logger/README.md (hunk 22 above)
Comment (Setogit) [inline packages/logger/README.md]: mentioned `@api`  thanks!

File: packages/logger/README.md (hunk 23 above)
Comment (Setogit) [inline packages/logger/README.md]: added `js` to all code blocks

File: packages/logger/README.md (hunk 24 above)
Comment (Setogit) [inline packages/logger/README.md]: yup, corrected.  thanks!

**Comment by Setogit** [comment]
@bajtos : incorporated all comments but `localhost:3000/openapi.json`

File: packages/logger/README.md [hunk 33]
```diff
@@ -0,0 +1,341 @@
...
+
+When I wanted to understand what's not there, I went to the spec and other options, but that rarely occurred.
+
+### [TypeScript](https://www.typescriptlang.org/) : Type System on ECMAScript 8
+
+1. Getting Started With TypeScript -- https://basarat.gitbooks.io/typescript/docs/getting-started.html 
+2. Declaration Files -- https://basarat.gitbooks.io/typescript/docs/types/ambient/d.ts.html
+3. Decorators -- https://www.typescriptlang.org/docs/handbook/decorators.html
+
+When you need to `npm install passport` don't forget to `npm install -S @types/passport
```
Comment (bajtos) [inline packages/logger/README.md]: Missing backtick at the end of this line.

File: packages/logger/README.md [hunk 34]
```diff
@@ -0,0 +1,341 @@
...
+```js
+class MyClass {
+  constructor(
+    @inject('defaultName') name: string
+  } {}
+}
+```
+In this case, the argument `name` is replaced with the value bound to the key `defaultName`.  @inject acquires the value bound to the `defaultName` key in the application's context, then assigns the value to `name` argument when the MyClass is instantiated.
+
+Who binds `defaultName` key to the value?  It depends.  You can do that by app.bind(`defaultName`).to('Ted Johnson').  You can `get` the value string by app.getSync('defaultName').
```
Comment (bajtos) [inline packages/logger/README.md]: `app.get('defaultName')` please

**Comment by bajtos** [comment]
The purpose of the logger work was to review our documentation and APIs to verify that an extension for logging can be built, which we have proved. I am closing this pull request as it's no longer needed.
